CMakeLists.txt

```diff
@@ -207,22 +207,40 @@ endif()
 
 # If available and wanted, use compiler cache.
 if(LIBCELLML_COMPILER_CACHE)
-  if(WIN32)
-    if(CLCACHE_EXE)
-      set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
-      execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
-                      RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
-      if(RESULT EQUAL 0)
-        set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
-      endif()
-    endif()
+  if(BUILDCACHE_EXE)
+    set(CMAKE_CXX_COMPILER_LAUNCHER ${BUILDCACHE_EXE})
   else()
-    if(CCACHE_EXE)
-      set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+    if(WIN32)
+      if(CLCACHE_EXE)
+        set(CLCACHEWRAPPER ${CMAKE_CURRENT_BINARY_DIR}/clcachewrapper)
+        execute_process(COMMAND ${CMAKE_C_COMPILER} /O2 /Fe${CLCACHEWRAPPER} ${CMAKE_CURRENT_SOURCE_DIR}/cmake/clcachewrapper.c
+                        RESULT_VARIABLE RESULT OUTPUT_QUIET ERROR_QUIET)
+        if(RESULT EQUAL 0)
+          set(CMAKE_CXX_COMPILER_LAUNCHER ${CLCACHEWRAPPER})
+        endif()
+      endif()
+    else()
+      if(CCACHE_EXE)
+        set(CMAKE_CXX_COMPILER_LAUNCHER ${CCACHE_EXE})
+      endif()
     endif()
   endif()
 endif()
 
+# On Windows, if we are using MSVC then replace some compiler flags.
+if(WIN32 AND MSVC)
+  # If we are treating warnings as errors then remove the /W3 option since it 
+  # will eventually be replaced with the /W4 and /WX options.
+  if(LIBCELLML_TREAT_WARNINGS_AS_ERRORS)
+    replace_compiler_flag("/W3" "")
+  endif()
+
+  # Replace the /Zi option (i.e. enable debugging information), if present, with
+  # the /Z7 option (i.e. enable old-style debugging information) otherwise
+  # caching (be it with buildcache or clcache) won't work.
+  replace_compiler_flag("/Z[iI]" "/Z7")
```

Comment:
I guess this should be wrapped with an "if using buildcache" ?

Reply:
Technically speaking, this is only needed when building using `buildcache` or `clcache`. This being said, it doesn't harm having whether we use a compiler cache or not. In fact, I would leave it as-is to be consistent on Windows with MSVC.

Reply:
I'm going to assume that I don't want to be using "old style debugging information" in Visual Studio just to enable something that I can't use.

Reply:
Done in commit ada6e1b0.